### ISSUE TITLE
Integrate Swedish realtime feeds

### ISSUE THREAD

**Issue by vkrause** [public-transport/transitous#125]
See https://www.trafiklab.se/api/.

**Comment by vkrause** [comment]
Right, that's the only option I see as well, similar to the situation in DE.

**Comment by 1Maxnet1** [comment]
> We could just add all of them and hope de-duplication is fast enough

Which kind of de-duplication is currently in place? Otherwise there are also existing tools to combine GTFS-RT feeds, which might be worth a look: https://github.com/mbta/concentrate

**Comment by vkrause** [comment]
> Which kind of de-duplication is currently in place?

We don't have any Transitous-specific deduplication so far, only what's already in MOTIS.

**Comment by felixguendling** [comment]
> We could just add all of them and hope deduplication is fast enough

With the latest release v0.12.5 it should be fast enough. At least before it was approx 50% of the import time, now it's gone from the profiler. So my hope would be that this approach would work on a larger scale (e.g. also in Germany).

Are there any blockers in MOTIS to try this? If I remember correctly, I read somewhere that Sweden uses the alternative method to identify trips via `route_id`, `direction`, etc. (maybe I'm wrong.. can't find it anymore) but I tried to check with the first feed from https://www.trafiklab.se/api/ and didn't get any `unsupported_no_trip_id` in the statistics.

**Comment by 1Maxnet1** [comment]
> This looks a bit complicacted:
> 
>     * [GTFS Sverige 2](https://www.trafiklab.se/api/trafiklab-apis/gtfs-sverige-2/) has all timetables but no realtime data
> 
>     * [GTFS Sweden 3](https://www.trafiklab.se/api/trafiklab-apis/gtfs-sweden/) has some timetables but realtime data
> 
>     * [GTFS Regional](https://www.trafiklab.se/api/trafiklab-apis/gtfs-regional/) has only regional connections but realtime data.
> 
> 
> We could just add all of them and hope deduplication is fast enough

There is also the NeTeX feed available, which according to their web page has the highest data quality and Realtime Available, but not complete coverage. To access the realtime feeds or the Netex datasets, one needs to register an API Key. Does anyone already have one we can use? Otherwise I might look into it in the next weeks and register an API Key for it.

**Comment by Brickblock1** [comment]
I took the time to write down most of what I know about these feeds feel free to ask any questions which may arise. 😊

# Feeds

GTFS 2 is the oldest and the one transitous is currently using.

NeTEx/GTFS Regional and Sweden are the same in terms of their backend and ids are aligned but prefixed ``SE:[Codespace]:[Type]:`` in NeTEx. It used to be that not all operators were avaliable in this system but that's no longer the case from what I can tell. 

The Sweden feeds combine all of the different regional feeds. They're sometimes cleaned of duplicates and certain lines or operators in the regional feeds are removed despite not existing elsewhere. Agency names are currently really messed up AB SL Kundtjänt being used instead of expected Storstockholms Lokaltrafik.

The Regional feeds are per datasource and contain slightly more info than the Sweden one. Ids are not shared with the Sweden feed.

There's also the Stops and Sales NeTEx datasets.

The stops dataset contains all the stops of NeTEx Sweden and their regional ids, can be usefull for finding issues in the merging (there are many), but lacks timetables so is of no use in this case.

The sales dataset is more interesting it's currently more related to GTFS 2 than the others it's the most incomplete only containing agencies whom are sold through Resplus/Samtrafiken Access. It does however have data not avaliable in the other feeds, such as product category, vehicle code etc. These are however expressed in keyLists.

All NeTEx feeds use the "Nordic" NeTEx profile which is very different from EPIP.

# Sources

## Noptis

Regional public transport authorities. Generally very good conversion into both NeTEx and GTFS.
Shapes, Quay information

## NeTEx 

### SJ, Snälltåget
Lines are train numbers for some reason. 
Works well with NeTEx but runs into issues with GTFS.
No shapes, No Quay information

### Vy

Lines are logical have PublicCode and Name. PublicCode is sometimes old timetable number but unknown to travler/public.
No shapes, No Quay information

## Trafikverket MPK

Regional rail operators: SL, Mälartåg, Västtåg, Öresundståg, Pågatåg, Tåg i Bergslagen, Norrtåg, Krösatåg, Östgötapendeln. In Regional as Trafikverket RDB.

One line per operator with name and PublicCode matching the operator. One SJ train number also here which is probably for testing but annoying since it gets dublicated. No shapes, No Quay information

## Trafikdataportalen

Smaller operators, data is often bad due to limitations in the tool and probably lazyness. Can't build proper StopPlaces which can cause issues when aggregating. Every trip gets it's own shape/JournayPattern.
No Shapes, Quay information sometimes


# Stops

## General

Stops in Sweden exist on 5 levels:
- Resplus/Samtrifiken Meta stop (Sales and GTFS 3 Sweden)
- Rikshållplats (Netex sales, stops, GTFS 2 and GTFS 3 Sweden)
- NeTEx GroupOfStopPlaces / Noptis Site (NeTEx Regional)
- GTFS Station / NeTEx StopPlace / Noptis Stop Area
- GTFS Platform / NeTEx Quay / Noptis Stop point

Meta stops are groups of Rikshållplatser used so that you can type Stockholm into your journeyplanner and not have to worry about it being Stockholm C, City or T-Centralen. 

Rikshållplatser are basically just stations but two stations can share a rikshållplats so they're technically a level above. The Rikshållplatsnumber is used for sales. NeTEx Sales and GTFS 2 have StopPlaces/stops be equivalent to rikshållplatser.

NeTEx GroupOfStopPlaces / Noptis Site are much like meta stops just for the regional feeds. SL uses these as the only searchable element in their journeyplanner, some sites therefore only contain 1 StopPlace. Many operators don't use these. Only avaliable in NeTEx Regional.

## Feed info

GTFS 2 only has rikshållplatser you can thereby not see what quay your trip departs from. Aggregation is good but not of much use without quays.

GTFS/NeTEx Sweden has StopPlaces and Quays but trains (Except SJ, Snälltåget, Vy, Tågab) are taken from Trafikverket directly so have no quay information. Vy might have quays but the rest of the trains do not. Some small operators may not separate their quays properly so will be wrong. The aggregation is often wrong. GTFS Sweden has Meta stops and rikshållplatser in areas.txt. 

Regional feeds have StopPlaces and Quays. Trains may have quays specified but it depends on the datasource Stockholm and Malmö does Gothenburg probably also does. Trafikverket RDB will contain all regional train in the way they are provided in the Sweden feed. It is the only source of Mälartåg, Norrtåg and TiB. NeTEx Regional provides GroupOfStopPlaces for some datasources.

NeTEx sales has only rikshållplatser and meta stops. Aggregation is good but not of much use without quays.

# Trips

GTFS 2: Train number in ``trip_short_name``, No shapes, Headsign is final stop name.

GTFS Sweden / Regional: Should but doesn't have train number in ``trip_short_name``, avaliable in ``samtrafiken_internal_trip_number`` often same as ``route_short_name``, Shapes of varying quality. Replacement busses often show up as trains.

NeTEx: Train number in ``TrainNumber``. Train number may change along route, Snälltåget does not do this (but should). Shapes of varying quality. ServiceFacilitySet avaliable for NeTEx and Trafikverket sources.

NeTEx sales extensions: product and vehicle codes avaliable. [Codelist](https://samtrafiken.atlassian.net/wiki/spaces/SamA/pages/3735617539/Codelist+Attributes+service+properties). County border crossings included. No shapes
# Routes

GTFS 2: ``route_short_name`` for trains is Train number if avaliable else ``route_long_name`` is operator / product. GTFS extended route types. SL Lokalbanor correctly trains.

GTFS Sweden / Regional: ``route_short_name`` for trains is Train number (Except Snälltåget, VR, Vy, Tågab, SL). ``route_long_name`` only shows if it differs from ``route_short_name``. A new route is created when the destination display public code differs from the ``route_short_name``, this makes it so the ``route_long_name`` becomes the original ``route_short_name`` instead of being empty. GTFS extended route types. SL Lokalbanor incorrectly trams.

NeTEx: Line Public code and name correct except for (VR, Vy, Tågab). ``DestinationDisplay/PublicCode`` Should be used instead of ``Line or ServiceJourney/PublicCode`` when avaliable. No Transport subModes but NeTEx sales has ProductCodes which can be used to fill the gaps. Noptis sources also have ``GroupOfLines/Name`` which can also be used.

# Realtime

Realtime is only avaliable for GTFS/NeTEx Regional and GTFS Sweden. There is realtime data for Flixbus despite what the trafiklab.se will tell you. Linking is very good for Noptis data but somewhat worse for the PTAs providing realtime via SIRI. Some trains have realtime in the regional feeds, these updates do not seem to be removed in the Sweden feed despite the trains being replaced. Some messages in Stockholm are internal and not meant for public display / only meant to be shown in the departure board, these can be filtered out from the SIRI but not from GTFS-RT.

# Thoughts

 As much as I prefer working with NeTEx I think it might be worth just doing GTFS Regional for the operators with realtime for now. The optimal soloution would probably be NeTEx Sweden + GTFS-RT Sweden with a few tweaks to VR and Vy Trains.
